### PR TITLE
Headhumpers

### PR DESCRIPTION
**Pull request by gurfan**
Scary headcrabs from half-life! Currently only obtainable through adminspawn.

Spritework and ideas from Vonadi.

They function similarly to facehuggers but with a few noticable differences:

- Xenomorphs can't pick them up or interact with them, they don't give a shit about eachother.
- Instead of slowly hopping one tile at a time towards a target, headcrabs will leap a couple tiles every time they move!
- Headcrabs take up the head slot instead of the mask slot (obviously)
- If the victim is wearing a regular hat, theres a 20% chance that it will deflect the headcrab. If the victim is wearing a rigsuit helmet, that chance is bumped up to 85%.
- Once a headcrab latches on, you won't be paralyzed like you would be from a facehugger. Instead, you'll be slowed momentarily.
- When a headcrab is latched onto you, you'll rapidly take brute damage. Start struggling!

You can get a headcrab off in a couple of ways:
- Using the resist verb or clicking on the headcrab in your inventory will cause you to attempt to remove it after 3 seconds, with a 50% chance of failing.
- If somebody else clicks on you in help intent (like a hug), they'll remove the headcrab after 2 seconds, with no chance of failing.

If you die or fall into critical condition when a headcrab is latched onto you, you'll shake violently for about short time and transform into a zombie. The headcrab can still be removed by another person before you're zombified.

Headcrab zombies are player controlled, and are mostly identical to normal zombies. Different zombie types as seen in hl2 might come in a later PR.
A few notable differences are:
- Any words spoken by the zombies will come out backwards.
- These zombies have slightly increased health compared to normal ones (hp 50 -> 75)
- When killed, they wont regenerate. Instead, the body of the victim is left behind, but the headcrab will disconnect from the victim.
- If the headcrab was still alive when the zombie was created, they'll still be alive when the zombie dies! Be careful!
- When the zombie is killed and the headcrab detaches itself, the head will have a 33% chance of exploding, leaving behind the victims brain.

Photo Gallery (Spritework by Vonadi)

humperfix
yeet
deadcrab
FUG
zombo

[content]

:cl:
 * rscadd: Headcrabs have finally snuck their way into the solar system. Thankfully, the station is still safe from them, at least for now.

### CODE REVIEW

File: code/modules/mob/living/carbon/alien/special/facehugger.dm [hunk 1]
```diff
@@ -413,11 +416,170 @@
...
+
+	while(target && target.head == src && !target.isDead() && !target.isInCrit())	//If they're still alive chew at their fuggin skull
+		playsound(src, 'sound/weapons/bite.ogg', 50, 1)
+		target.apply_damage(10, BRUTE, LIMB_HEAD)
+		sleep(30)
+
+	if(target && target.head == src && (target.isDead() || target.isInCrit()))	//Once they die, start the zombification.
+		visible_message("<span class='danger'>[target.real_name] begins to shake and convulse violently!</span>")
+		to_chat(target, "<span class='sinister'>You feel your consciousness slipping away...</span>")
+		target.Jitter(1000)
```
Comment (boy2mantwicethefam) [inline code/modules/mob/living/carbon/alien/special/facehugger.dm]: ah yes, 2000 seconds of jittering

File: code/modules/mob/living/simple_animal/hostile/necro.dm [hunk 2]
```diff
@@ -576,3 +576,54 @@
 				set_light(1, 2, "#5dca31")
 
 #undef RAD_COST
+
+
+/////////////////////////////////////////////////////// 
+///////////////// HEADCRAB ZOMBIES //////////////////// 
+///////////////////////////////////////////////////////
+
+/mob/living/simple_animal/hostile/necro/zombie/headcrab //Not very useful
```
Comment (boy2mantwicethefam) [inline code/modules/mob/living/simple_animal/hostile/necro.dm:585]: because they're a child of zombies there's a possibility that they inherit zombie behavior including eating meat, and player zombies can turn into other kinds of zombies. Investigate if this holds true for headcrab zombies

Comment (boy2mantwicethefam) [reply code/modules/mob/living/simple_animal/hostile/necro.dm:585]: Nevermind, they got can_evolve set to false. Look into...

File: code/modules/mob/living/simple_animal/hostile/necro.dm [hunk 3]
```diff
@@ -576,3 +576,54 @@
...
+	icon_state = "zombie_turned" //Looks almost not unlike just a naked guy to potentially catch others off guard
+	icon_living = "zombie_turned"
+	icon_dead = "zombie_turned"
+	desc = "A reanimated corpse that looks like it has seen better days. This one still appears quite human."
+	maxHealth = 75
+	health = 75
+	can_evolve = FALSE
+	canRegenerate = 1
```
Comment (boy2mantwicethefam) [inline code/modules/mob/living/simple_animal/hostile/necro.dm]: canRegenerate though. If my guess is right that'd mean they rise up after a while after death

File: code/modules/mob/living/simple_animal/hostile/necro.dm [hunk 4]
```diff
@@ -576,3 +576,54 @@
...
+	harm_intent_damage = 20
+	melee_damage_lower = 15
+	melee_damage_upper = 2
+	var/mob/living/carbon/human/host //Whoever the zombie was previously, kept in a reference to potentially bring back
+
+/mob/living/simple_animal/hostile/necro/zombie/headcrab/New()
```
Comment (boy2mantwicethefam) [inline code/modules/mob/living/simple_animal/hostile/necro.dm]: Unnecessary New()

File: code/modules/mob/living/simple_animal/hostile/necro.dm [hunk 5]
```diff
@@ -576,3 +576,54 @@
...
+		mind.transfer_to(host)
+		var/mob/dead/observer/ghost = get_ghost_from_mind(mind)
+		if(ghost && ghost.can_reenter_corpse)
+			key = mind.key
+		if(prob(33))	//33% chance to blow up their fucking head
+			var/datum/organ/external/head/head_organ = host.get_organ(LIMB_HEAD)
+			head_organ.explode()
+		else
+			visible_message("<span class='danger'>The headcrab releases it's grasp from [src]!</span>")
+		host = null
```
Comment (boy2mantwicethefam) [inline code/modules/mob/living/simple_animal/hostile/necro.dm:620]: You might want to qdel the host before setting this to null

File: code/modules/mob/living/carbon/alien/special/facehugger.dm [hunk 6]
```diff
@@ -413,11 +416,170 @@
...
+			else
+				H.visible_message("<span class='danger'>\The [src] bounces off of \the [head_protection]!</span>")
+				if(prob(CHANCE_TO_DIE_AFTER_HEAD_DENIED) && !sterile)
+					death()
+					return
+				else
+					GoIdle(TIME_IDLE_AFTER_HEAD_DENIED)
+					return
+
+	if(iscarbon(L))
```
Comment (boy2mantwicethefam) [inline code/modules/mob/living/carbon/alien/special/facehugger.dm]: I don't think this is a good idea, there's lots of kinds of carbons and it'll be pretty janky. Best to restrict this to humans for now

**Comment by DrSnips** [comment]
Player zombies are already stupid strong at their current stats. They don't need the buffs.

Hits to the zombie's head should mirror damage to the headcrab giving people at least some chance to stop it. You're going to have to take 5x 15 damage hits to kill the zombie before you can even hope to deal with the headcrab seems like a problem. Not to mention this isn't a normal mindless zombie that will willingly take these hits, player zombies actually can dodge and dance.

File: code/modules/mob/living/carbon/alien/special/facehugger.dm [hunk 7]
```diff
@@ -413,11 +416,170 @@
...
 	return sterile
+
+
+
+
+//////////////////////////////////////////////////// 
+////////////////  HEADCRABS  /////////////////////// 
+////////////////////////////////////////////////////
+
+/obj/item/clothing/mask/facehugger/headcrab
```
Comment (boy2mantwicethefam) [inline code/modules/mob/living/carbon/alien/special/facehugger.dm:438]: uhhh, did you add a Process() anywhere to handle headcrab behavior?

File: code/modules/mob/living/carbon/alien/special/facehugger.dm [hunk 8]
```diff
@@ -413,11 +416,170 @@
...
+	return FALSE
+
+/obj/item/clothing/mask/facehugger/headcrab/proc/Assimilate(mob/living/L)
+	if(!ishuman(L))
+		return
+	var/mob/living/carbon/human/target = L
+	if(!target || target.head != src) //was taken off or something
+		return
+
+	while(target && target.head == src && !target.isDead() && !target.isInCrit())	//If they're still alive chew at their fuggin skull
```
Comment (boy2mantwicethefam) [inline code/modules/mob/living/carbon/alien/special/facehugger.dm:617]: "while" happens very very fast IIRC, I know about the sleep(30) though I'd recommend moving this to a Process() along with a bunch of stuff

File: code/modules/mob/living/simple_animal/hostile/necro.dm (hunk 3 above)
Comment (gurfan) [reply code/modules/mob/living/simple_animal/hostile/necro.dm]: Oops, meant to set that to zero

File: code/modules/mob/living/simple_animal/hostile/necro.dm (hunk 5 above)
Comment (ghost) [reply code/modules/mob/living/simple_animal/hostile/necro.dm:620]: No you don't, setting host to null here is unnecessary to begin with. You only need to clear refs to the datum that you're deleting, not the ones inside of the datum you're deleting.

File: code/modules/mob/living/carbon/alien/special/facehugger.dm (hunk 7 above)
Comment (ghost) [reply code/modules/mob/living/carbon/alien/special/facehugger.dm:438]: Why would you need a process? Nothing is changing state dynamically, everything seems to be safely handled when the headcrab is equipped/removed.

Comment (gurfan) [reply code/modules/mob/living/carbon/alien/special/facehugger.dm:438]: It inherits process() from facehuggers, but I should move things like the damage while latched on to a process(), unless the while() works fine for that...?

File: code/modules/mob/living/carbon/alien/special/facehugger.dm (hunk 8 above)
Comment (ghost) [reply code/modules/mob/living/carbon/alien/special/facehugger.dm:617]: It sleeps for 3 seconds at the end of the proc.

**Comment by juanototo2** [comment]
The damage was fine. These zombies need a headcrab for EVERY human they want infected, unlike normal zombies which can effortlessly multiply. That, and they only get one life each.

**Comment by tuypo1** [comment]
I do agree shooting the head should probably also damage the crap, although it might be too unbalanced if it ever comes time to try implementing these outside adminspawn

**Comment by gurfan** [comment]
I still have some things that I am working on, if anyone else has suggestions please let me know

**Comment by DeityLink** [comment]
Now to recolor the gravity well gun and recode it into an actual gravity gun

**Comment by gurfan** [comment]
Tweaks:

- When the headcrab zombie is killed, the headcrab will immediately leap away from the nearest person in an attempt to escape. Watch out though, they'll quickly start chasing again after a few seconds.
- Headcrabs will chase from a slightly longer range. Their jump distance remains the same at about 3 tiles.

**Comment by gurfan** [comment]
- headcrabs will try to jump away when you resist them off
- bugfixes

File: code/modules/mob/living/carbon/alien/special/facehugger.dm [hunk 9]
```diff
@@ -199,9 +202,11 @@
 	return FALSE
 
 /obj/item/clothing/mask/facehugger/HasProximity(atom/movable/AM as mob|obj)
-	if(isliving(AM))
+	if(ishuman(AM))
 		if(CanHug(AM, src))
-			return Attach(AM)
+			var/mob/living/carbon/human/H
+			if(!H.isUnconscious())
+				Attach(H)
```
Comment (Zth--) [inline code/modules/mob/living/carbon/alien/special/facehugger.dm]: why did you remove the return? Now this proc always returns false

File: code/modules/mob/living/carbon/alien/special/facehugger.dm [hunk 10]
```diff
@@ -413,11 +420,211 @@
...
+		..()
+
+/obj/item/clothing/mask/facehugger/headcrab/findtarget()
+	if(!real)
+		return
+	target = null
+	for(var/mob/living/carbon/human/T in hearers(src,6))
+		if(!CanHug(T, src))
+			continue
+		if(T && !T.isUnconscious())
```
Comment (Zth--) [inline code/modules/mob/living/carbon/alien/special/facehugger.dm]:
```suggestion
		if(!T?.isUnconscious())
```

File: code/modules/mob/living/carbon/alien/special/facehugger.dm [hunk 11]
```diff
@@ -413,11 +420,211 @@
...
+				sleep(50)
+				GoActive()
+			else
+				throw_at(target, 3, 1)
+				if(dist == 0)
+					if(CanHug(target, src) && isturf(target.loc)) //Fix for hugging through mechs and closets
+						Attach(target)
+						return
+
+/obj/item/clothing/mask/facehugger/headcrab/Attach(mob/living/L)
```
Comment (Zth--) [inline code/modules/mob/living/carbon/alien/special/facehugger.dm:537]: this proc doesn't seem to ever return true

File: code/modules/mob/living/simple_animal/hostile/necro.dm [hunk 12]
```diff
@@ -576,3 +576,56 @@
...
+			mind.transfer_to(host)
+			var/mob/dead/observer/ghost = get_ghost_from_mind(mind)
+			if(ghost && ghost.can_reenter_corpse)
+				key = mind.key
+		if(prob(33))	//33% chance to blow up their fucking head
+			var/datum/organ/external/head/head_organ = host.get_organ(LIMB_HEAD)
+			head_organ.explode()
+		else
+			visible_message("<span class='danger'>The headcrab releases it's grasp from [src]!</span>")
+		crab?.escaping = 1
```
Comment (Zth--) [inline code/modules/mob/living/simple_animal/hostile/necro.dm:617]: why does this line use the `?.` operator and the one below doesn't?

File: code/modules/mob/living/carbon/alien/special/facehugger.dm [hunk 13]
```diff
@@ -168,7 +168,10 @@
 		if(CONSCIOUS)
 			to_chat(user, "<span class='danger'>\The [src] seems active.</span>")
 	if (sterile)
-		to_chat(user, "<span class='danger'>It looks like \the [src]'s proboscis has been removed.</span>")
+		if(istype(src, /obj/item/clothing/mask/facehugger/headcrab))
```
Comment (DamianX) [inline code/modules/mob/living/carbon/alien/special/facehugger.dm]: Implement `/obj/item/clothing/mask/facehugger/headcrab` instead of typechecking `src`

File: code/modules/mob/living/carbon/alien/special/facehugger.dm [hunk 14]
```diff
@@ -199,9 +202,11 @@
 	return FALSE
 
 /obj/item/clothing/mask/facehugger/HasProximity(atom/movable/AM as mob|obj)
-	if(isliving(AM))
+	if(ishuman(AM))
 		if(CanHug(AM, src))
-			return Attach(AM)
+			var/mob/living/carbon/human/H
```
Comment (DamianX) [inline code/modules/mob/living/carbon/alien/special/facehugger.dm]: this is always null, will runtime.
```suggestion
			var/mob/living/carbon/human/H = AM
```

File: code/modules/mob/living/carbon/alien/special/facehugger.dm [hunk 15]
```diff
@@ -413,11 +420,211 @@
...
+
+/obj/item/clothing/mask/facehugger/headcrab/equipped(mob/living/carbon/human/H)
+	if(stat == CONSCIOUS)
+		Assimilate(H)
+	else if(stat == UNCONSCIOUS)
+		GoActive()
+		Assimilate(H)
+
+/obj/item/clothing/mask/facehugger/headcrab/attack_hand(mob/user)
+	var/mob/living/carbon/human/target = user
```
Comment (DamianX) [inline code/modules/mob/living/carbon/alien/special/facehugger.dm]: you need to typecheck `target` here, humans are not the only mob that can `attack_hand`

File: code/modules/mob/living/carbon/alien/special/facehugger.dm [hunk 16]
```diff
@@ -413,11 +420,211 @@
...
+		..()
+
+/obj/item/clothing/mask/facehugger/headcrab/findtarget()
+	if(!real)
+		return
+	target = null
+	for(var/mob/living/carbon/human/T in hearers(src,6))
+		if(!CanHug(T, src))
+			continue
+		if(!T?.isUnconscious())
```
Comment (DamianX) [inline code/modules/mob/living/carbon/alien/special/facehugger.dm]:
```suggestion
		if(!T.isUnconscious())
```
`T` should never be null here

File: code/modules/mob/living/carbon/alien/special/facehugger.dm [hunk 17]
```diff
@@ -413,11 +420,211 @@
...
+			if(get_dist(loc, T.loc) <= 6)
```
Comment (DamianX) [inline code/modules/mob/living/carbon/alien/special/facehugger.dm]: how can someone be in `hearers(src, 6)` but fail this check?

File: code/modules/mob/living/carbon/alien/special/facehugger.dm [hunk 18]
```diff
@@ -413,11 +420,211 @@
...
+		healthcheck()
+
+
+/obj/item/clothing/mask/facehugger/headcrab/followtarget()
+	if(!real)
+		return // Why are you trying to path stupid toy
+	if(!target)
+		findtarget()
+		return
+	if(loc && isturf(loc) && !attached && !stat && nextwalk <= world.time)
```
Comment (DamianX) [inline code/modules/mob/living/carbon/alien/special/facehugger.dm]:
```suggestion
	if(isturf(loc) && !attached && !stat && nextwalk <= world.time)
```

File: code/modules/mob/living/carbon/alien/special/facehugger.dm [hunk 19]
```diff
@@ -413,11 +420,211 @@
...
+	GoIdle(TIME_IDLE_AFTER_ATTACH_DENIED) 
+
+
+	return FALSE
+
+/obj/item/clothing/mask/facehugger/headcrab/proc/Assimilate(mob/living/L)
+	if(!ishuman(L))
+		return
+	var/mob/living/carbon/human/target = L
+	if(!target || target.head != src) //was taken off or something
```
Comment (DamianX) [inline code/modules/mob/living/carbon/alien/special/facehugger.dm]:
```suggestion
	if(target.head != src) //was taken off or something
```
target can never be null here

File: code/modules/mob/living/carbon/human/human_attackhand.dm [hunk 20]
```diff
@@ -233,7 +233,14 @@
...
 	switch(M.a_intent)
 		if(I_HELP)
-			if(health >= config.health_threshold_crit)
+			if(istype(head, /obj/item/clothing/mask/facehugger/headcrab))
+				var/obj/item/clothing/mask/facehugger/headcrab/crab = get_item_by_slot(slot_head)
+				if(do_after(M, src, 2 SECONDS))
+					if(head == crab)
+						drop_from_inventory(crab)
+						crab.GoIdle(15 SECONDS) 
+						visible_message("[M.real_name] pulls the headcrab off of [src.real_name]'s head!")
```
Comment (DamianX) [inline code/modules/mob/living/carbon/human/human_attackhand.dm]:
```suggestion
						visible_message("[M] pulls the headcrab off of [src]'s head!")
```

**Review by DamianX — COMMENTED** [review]
what a mess

File: code/modules/mob/living/carbon/alien/special/facehugger.dm [hunk 21]
```diff
@@ -62,9 +63,7 @@
 		if(!CanHug(T, src))
 			continue
 		if(T && (!T.isUnconscious() ) )
```
Comment (Zth--) [inline code/modules/mob/living/carbon/alien/special/facehugger.dm:65]:
```suggestion
		if(!T.isUnconscious())
```

T cannot ever be null here
